[Completion] Better handle merging of lookup base types

hamishknight opened the pull request:
For unresolved member completion, we were preferring the more general type, when we ought to be preferring the more specific type. Additionally, for both unresolved member and postfix completion we were opening archetypes, which doesn't work as expected since we don't compare requirements. Factor out the logic that deals with merging base types for lookup, and have it prefer either the subtype, or the optional type in the case of optional promotion.

rdar://126168123

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

rintaro commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -60,11 +60,20 @@ namespace swift {
+  /// means preferring a subtype over a supertype, but can also e.g prefer an
+  /// optional over a non-optional. If the two types are incompatible, null is
+  /// returned.
+  Type tryMergeBaseTypeForLookup(Type ty1, Type ty2, DeclContext *dc);
```
Is this is usable other than code completion? If not, maybe `tryMergeBaseTypeForCompletionLookup`

ahoppen reviewed (approved):
Very nice

hamishknight committed:
[Completion] Better handle merging of lookup base types

For unresolved member completion, we were preferring
the more general type, when we ought to be preferring
the more specific type. Additionally, for both
unresolved member and postfix completion we were
opening archetypes, which doesn't work as expected
since we don't compare requirements. Factor out
the logic that deals with merging base types for
lookup, and have it prefer either the subtype, or
the optional type in the case of optional promotion.

rdar://126168123

hamishknight commented:
@swift-ci please smoke test
